File: Aditya Vardhan Singh/Other Problems/Day 15/Find pairs with given sum in doubly linked list/Python/Find pairs with given sum in doubly linked list.py
from typing import Optional


from typing import List

"""

Definition for singly Link List Node
class Node:
    def __init__(self,x):
        self.data=x
        self.next=None
        self.prev=None

You can also use the following for printing the link list.
displayList(node)
"""

class Solution:
    def findPairsWithGivenSum(self, target : int, head : Optional['Node']) -> List[List[int]]:
        # code here
        ans = []
        if head.next is None:
            return ans
        
        p1 = head
        p2 = head
        
        while p2.next:
            p2 = p2.next
        
        while p1 is not p2 and p2.next is not p1:
            d1, d2 = p1.data, p2.data
            
            if (d1 + d2) == target:
                ans.append([d1, d2])
                p1 = p1.next
                p2 = p2.prev
            elif (d1 + d2) < target:
                p1 = p1.next
            else:
                p2 = p2.prev
        
        return ans

File: Aditya Vardhan Singh/Other Problems/Day 15/Find pairs with given sum in doubly linked list/C++/Find pairs with given sum in doubly linked list.cpp
//User function Template for C++

/* Doubly linked list node class
class Node
{
public:
    int data;
    Node *next, *prev;
    Node(int val) : data(val), next(NULL), prev(NULL)
    {
    }
};
*/

class Solution
{
public:
    vector<pair<int, int>> findPairsWithGivenSum(Node *head, int target)
    {
        // code here
        if (head->next == nullptr) {
            return {{}};
        }
        
        vector<pair<int, int>> ans;
        Node *front = head;
        Node *back = head;
        
        while (back->next) { back = back->next; }
        
        while (front != back && back->next != front) {
            int d1 = front->data;
            int d2 = back->data;
            
            if (d1 + d2 == target) {
                ans.push_back( {d1, d2} );
                front = front->next;
                back = back->prev;
            } else if (d1 + d2 < target) {
                front = front->next;
            } else {
                back = back->prev;
            }
        }
        
        return ans;
    }
};